jobs/webcompat-kb/webcompat_kb/main.py

```diff
@@ -590,30 +671,140 @@ def filter_bug_history_changes(self, history):
                         "changes": relevant_changes,
                     }
                     filtered_changes.append(filtered_record)
+                    bug_ids.add(bug_history["id"])
 
             if filtered_changes:
-                res.extend(filtered_changes)
+                result.extend(filtered_changes)
+
+        return result, bug_ids
+
+    def filter_relevant_history(self, updated_history):
+        only_unsaved_changes = []
+        result, bug_ids = self.extract_relevant_changes(updated_history)
 
-        return res
+        if result:
+            only_unsaved_changes = self.filter_only_unsaved_changes(result, bug_ids)
+
+        return only_unsaved_changes
 
     def get_bugs_updated_since_last_import(self, all_bugs, last_import_time):
         return [
             bug["id"]
             for bug in all_bugs
-            if datetime.strptime(bug["last_change_time"], "%Y-%m-%dT%H:%M:%SZ").replace(
-                tzinfo=timezone.utc
-            )
-            > last_import_time
+            if parse_datetime_str(bug["last_change_time"]) > last_import_time
         ]
 
-    def fetch_update_history(self, all_bugs):
-        last_import_time = self.get_last_import_datetime()
+    def get_imported_ids(self):
+        query = f"""
+                SELECT number
+                FROM `{self.bq_dataset_id}.bugzilla_bugs`
+            """
+        res = self.client.query(query).result()
+        rows = list(res)
+
+        imported_ids = {bug["number"] for bug in rows}
+
+        return imported_ids
+
+    def create_keyword_map(self, history):
+        keyword_history = {}
+        for record in history:
+            bug_id = record["number"]
+            for change in record["changes"]:
+                if "keywords" in change["field_name"]:
+                    if bug_id not in keyword_history:
+                        keyword_history[bug_id] = {"added": set(), "removed": set()}
+                    if change["added"]:
+                        keyword_history[bug_id]["added"].update(
+                            change["added"].split(", ")
+                        )
+                    if change["removed"]:
+                        keyword_history[bug_id]["removed"].update(
+                            change["removed"].split(", ")
+                        )
+        return keyword_history
+
+    def create_synthetic_history(self, bugs, history):
+        bug_ids = set()
+        keyword_history = self.create_keyword_map(history)
```

Comment:
Unless we're using this somewhere else, it seems like what we want is a per-bug set of keywords for which there's either no added/removed entry or a removed entry before any added entry. In particular it seems like this approach is going to miss case where a keyword is added on creation, then removed, then readded.